File: 53 Maximum Subarray/53.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        if len(nums) == 1:
            return nums[0]
        preSum = 0
        globalMax = -float('inf')
        for i in range(len(nums)):
            preSum = max(preSum + nums[i], nums[i])
            globalMax = max(globalMax, preSum)
        return globalMax

File: 53 Maximum Subarray/53.java
class Solution {
    public int maxSubArray(int[] nums) {
        if (nums.length == 1) {
            return nums[0];
        }
        int preSum = 0;
        int globalMax = Integer.MIN_VALUE;
        for (int i = 0; i < nums.length; ++i) {
            preSum = Math.max(preSum + nums[i], nums[i]);
            globalMax = Math.max(globalMax, preSum);
        }
        return globalMax;
    }
    // Time = O(n)
    // Space = O(1)
}